[Draft] Add CFG side panel

kingbri1 opened the pull request:
This proposed PR adds CFG for the following cases:
- Per-chat
- Per-character
- Global

Doing so adds more flexibility to CFG as a whole. If positive prompts become supported, add those as well. Currently only works with ooba UI and NOT NovelAI due to me not having an API key.

Replacement for #889

kingbri1 committed:
Extensions: Add global CFG support

CFG is bundled as an extension which is more flexible in terms of
storing settings and consolidating code.

Global CFG is currently supported and the old power user setting
is auto-migrated.

Signed-off-by: kingbri <bdashore3@proton.me>

TheBestReguile commented:
Have you experimented with this?  How does it perform/what prompts are you having luck with?

Is it possible to add an option to do something like:

Take the entire prompt that ST would have generated, including all the chat messages up to that point.
Append an instruction to it - kind of like how system messages work - something like "Respond with a short and terrible response"
Pass that whole thing, with all the context and everything, into the result?

I'd be curious to see how that performs vs a "type in a thing into CFG", since you'd be biasing against a "real" outcome.

kingbri1 committed:
CFG: Add per-character support

Adds per-character CFG as a drop-in replacement for global CFG. If
the use character CFG checkbox isn't checked, the global one will be
used.

Signed-off-by: kingbri <bdashore3@proton.me>

kingbri1 commented:
~~Note 2: This PR is NOT planned to add group chat support. That will have to be added at a later time.~~

kingbri1 committed:
CFG: Add per-chat CFG and fixes

Per-chat CFG applies a CFG setting per-chat only rather than character
or globally. This overrides all other CFG settings (this will be changed).

Also add fixes to remove character CFG entries properly and not to
apply CFG if the scale is 1 as that won't do anything to generation.

Signed-off-by: kingbri <bdashore3@proton.me>

kingbri1 committed:
CFG: Add ability to combine negative prompts

This allows for flexibility with global, character, and chat negative
prompts. Combining prompts is very useful for users who want to maintain
a set of global negatives and then add extra layers on top of that. The
ordering is chat -> character -> global tags due to the specificity of
each.

The guidance scale follows the cascade of chat -> character -> global
due to being one number that is set when CFG is fired. If the guidance
scale is 1, nothing happens.

Signed-off-by: kingbri <bdashore3@proton.me>

kingbri1 committed:
CFG: Add groupchat support

Groupchats have their CFG defined a little differently. Chat and
global CFGs are publicly accessible within the window. As a compromise,
individual character CFGs can be injected as part of each character's
prompt. However, these CFG settings need to be adjusted in the character's
individual chats which will carry over to the group.

In addition, make this character logic gated under a checkbox to
always prefer chat unless explicitly specified. Negative cascading
is still open, so individual character negatives can be included
at any time.

Signed-off-by: kingbri <bdashore3@proton.me>

kingbri1 commented:
Merge ready.
